src/System.IO.Compression.ZipFile/src/System/IO/Compression/ZipFile.cs

```diff
@@ -0,0 +1,546 @@
+        ///         <item>For entry names that do not contain characters outside the ASCII range,
+        ///         the language encoding flag (EFS) will not be set in the general purpose bit flag of the local file header,
+        ///         and the current system default code page (<c>Encoding.Default</c>) will be used to encode the entry names into bytes.</item>
+        ///     </list>
+        ///     <para>If <c>entryNameEncoding</c> is specified (<c>!= null</c>):</para>
+        ///     <list>
+        ///         <item>The specified <c>entryNameEncoding</c> will always be used to encode the entry names into bytes.
+        ///         The language encoding flag (EFS) in the general purpose bit flag of the local file header will be set if and only
+        ///         if the specified <c>entryNameEncoding</c> is a UTF-8 encoding.</item>
+        ///     </list>
+        ///     <para>Note that Unicode encodings other than UTF-8 may not be currently used for the <c>entryNameEncoding</c>,
+        ///     otherwise an <see cref="ArgumentException"/> is thrown.</para>
+        /// </param>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Reliability", "CA2000:Dispose objects before losing scope")]  // See comment in the body.
+        public static ZipArchive Open(String archiveFileName, ZipArchiveMode mode, Encoding entryNameEncoding)
+        {
+            // Relies on File.Open for checking of archiveFileName
+
+            FileMode fileMode;
+            FileAccess access;
+            FileShare fileShare;
+
+            switch (mode)
+            {
+                case ZipArchiveMode.Read:
+                    fileMode = FileMode.Open;
+                    access = FileAccess.Read;
+                    fileShare = FileShare.Read;
+                    break;
+
+                case ZipArchiveMode.Create:
+                    fileMode = FileMode.CreateNew;
+                    access = FileAccess.Write;
+                    fileShare = FileShare.None;
+                    break;
+
+                case ZipArchiveMode.Update:
+                    fileMode = FileMode.OpenOrCreate;
+                    access = FileAccess.ReadWrite;
+                    fileShare = FileShare.None;
+                    break;
+
+                default:
+                    throw new ArgumentOutOfRangeException("mode");
+            }
+
+            // Surpress CA2000: fs gets passed to the new ZipArchive, which stores it internally.
+            // The stream will then be owned by the archive and be disposed when the archive is disposed.        
+            // If the ctor completes without throwing, we know fs has been successfully stores in the archive;
+            // If the ctor throws, we need to close it here.
+
+            FileStream fs = null;
+
+            try
+            {
+                fs = File.Open(archiveFileName, fileMode, access, fileShare);
```

Comment:
Should this File.Open line be moved to before the try block?  Then the catch block can be simplified to just call Dispose and throw without first checking whether fs is null.